### PR TITLE
frontend: Create nav link for ingress

### PR DESCRIPTION
**Pull request by vyncent-t**
Related to issue #1268 

## Description
This PR addresses the enhancement requested in issue #1268 . The purpose was to transform the host in the ingress rules into a clickable link, thereby eliminating the need for users to manually copy and paste the URL to open the endpoint.

## Changes
image

- [x]  Functional links within the field Host
- [x]  Additional links for path field for specific routing
- [x] Created logic for determining url protocols based on ingress tls, defaults to http when not using tls

## Notes
- To validate this change, I created example pod, service, and host with the host set to bing. The changes work for formatted URL stings only, but will print '*' if needed and will not print any links if no host is written.
- When using tls and host is within fields, protocol is set to https, if not using tls or if it is not found, defaults to http

Using tls
image

Not using tls
image

### CODE REVIEW

File: frontend/src/components/common/LabelListItem.tsx [hunk 1]
```diff
@@ -6,7 +6,7 @@ export interface LabelListItemProps {
 }
 
 export default function LabelListItem(props: LabelListItemProps) {
-  const { labels } = props;
+  const { labels = [] } = props;
```
Comment (joaquimrocha) [inline frontend/src/components/common/LabelListItem.tsx:9]: This should be in a separate commit as it's an unrelated fix.

File: frontend/src/components/ingress/Details.tsx [hunk 2]
```diff
@@ -74,7 +75,8 @@ export default function IngressDetails() {
               columns={[
                 {
                   label: t('Host'),
-                  getter: (data: IngressRule) => data.host || '*',
+                  getter: (data: IngressRule) =>
+                    <Link routeName={`${data.host}`}>{data.host}</Link> || '*',
```
Comment (joaquimrocha) [inline frontend/src/components/ingress/Details.tsx:79]: The routeName is supposed to be a route's name from the object in router.tsx.
Also you need to validate whether the host is an actual URL before adding it as a link prop.

File: frontend/src/lib/k8s/ingress.ts [hunk 3]
```diff
@@ -84,7 +84,7 @@ class Ingress extends makeKubeObject<KubeIngress>('ingress') {
 
     const rules: IngressRule[] = [];
 
-    this.spec!.rules.forEach(({ http, host }) => {
+    this.spec!.rules?.forEach(({ http, host }) => {
```
Comment (joaquimrocha) [inline frontend/src/lib/k8s/ingress.ts:87]: This needs to be a separate commit.

**Review by joaquimrocha — CHANGES_REQUESTED** [review]

File: frontend/src/components/common/LabelListItem.tsx (hunk 1 above)
Comment (vyncent-t) [reply frontend/src/components/common/LabelListItem.tsx]: Put it in the fix branch!

File: frontend/src/lib/k8s/ingress.ts (hunk 3 above)
Comment (vyncent-t) [reply frontend/src/lib/k8s/ingress.ts]: Put it in the fix branch!

File: frontend/src/components/ingress/Details.tsx [hunk 4]
```diff
@@ -36,6 +37,29 @@ export default function IngressDetails() {
...
+      const formatURL = new URL(`http://${url}`);
+      if (formatURL) {
+        return (
+          <Link
+            href={`${formatURL.protocol}//${formatURL.hostname}`}
+          >{`${formatURL.protocol}//${formatURL.hostname}`}</Link>
+        );
+      }
+    } catch (e) {
+      throw new Error('Invalid URL');
```
Comment (joaquimrocha) [inline frontend/src/components/ingress/Details.tsx:59]: I think we should log the occurrence, though here it's just for debug purposes (console.debug("Ingress' host is not a valid URL; displaying it without a link.')) as it may fail for some reason. And it should return just the string IMO.

File: frontend/src/components/ingress/Details.tsx [hunk 5]
```diff
@@ -74,7 +98,7 @@ export default function IngressDetails() {
               columns={[
                 {
                   label: t('Host'),
-                  getter: (data: IngressRule) => data.host || '*',
+                  getter: (data: IngressRule) => isFormatURL(data.host || '*'),
```
Comment (joaquimrocha) [inline frontend/src/components/ingress/Details.tsx:101]: Should we include the path too? There's something to be decided about that here.

File: frontend/src/components/ingress/Details.tsx [hunk 6]
```diff
@@ -36,6 +37,29 @@ export default function IngressDetails() {
...
+  /** this function is used to format the url to be used in the Link component */
+  function isFormatURL(url: any) {
+    if (url === '*') {
+      return '*';
+    }
+    try {
+      /** https://developer.mozilla.org/en-US/docs/Web/API/URL/URL
+       * new URL() constructor creates a new URL object instance only if the string passed is a valid URL
+       * if the string is not a valid URL, then an error is thrown
+       * note: adding the protocol to the url string is required for the URL constructor to work because data.host is only the host name */
```
Comment (joaquimrocha) [inline frontend/src/components/ingress/Details.tsx:49]: Except for the last line (without the note prefix), I don't think the comment is needed as it's kind of the API doc.

File: frontend/src/components/ingress/Details.tsx [hunk 7]
```diff
@@ -36,6 +37,29 @@ export default function IngressDetails() {
     );
   }
 
+  /** this function is used to format the url to be used in the Link component */
+  function isFormatURL(url: any) {
+    if (url === '*') {
+      return '*';
+    }
```
Comment (joaquimrocha) [inline frontend/src/components/ingress/Details.tsx:44]: This could be removed since we'd catch it with the exception anyway, but exceptions should be avoided, so it's good! 👍

File: frontend/src/components/ingress/Details.tsx [hunk 8]
```diff
@@ -36,6 +37,29 @@ export default function IngressDetails() {
     );
   }
 
+  /** this function is used to format the url to be used in the Link component */
+  function isFormatURL(url: any) {
```
Comment (joaquimrocha) [inline frontend/src/components/ingress/Details.tsx:41]: The name of the function seems like it returns a boolean, but it returns a component. So it should be called something like LinkOrString or something like that.

**Review by joaquimrocha — CHANGES_REQUESTED** [review]
Needs a couple of fixes. Also, Ingresses may have an https link if they have TLS set up. So we should likely add that too. Ping me if you need more details.

File: frontend/src/components/ingress/Details.tsx (hunk 5 above)
Comment (vyncent-t) [reply frontend/src/components/ingress/Details.tsx:101]: Should we add the same path as the field below this?

**Comment by illume** [comment]
https://kubernetes.io/docs/concepts/services-networking/ingress/

File: frontend/src/components/ingress/Details.tsx (hunk 5 above)
Comment (illume) [reply frontend/src/components/ingress/Details.tsx]: I think you'd almost always want the path in the link. It doesn't seem to me very common that someone would want to visit just the host part.

**Comment by illume** [comment]
Missing from here is the 'prefix' information for the path. We should display this, because it's quite important... since it can change the meaning of it quite a lot.

File: frontend/src/components/ingress/Details.tsx (hunk 6 above)
Comment (vyncent-t) [reply frontend/src/components/ingress/Details.tsx]: Fixed it!

File: frontend/src/components/ingress/Details.tsx (hunk 7 above)
Comment (vyncent-t) [reply frontend/src/components/ingress/Details.tsx]: Fixed it!

**Comment by vyncent-t** [comment]
> Not exactly in scope for this... but...
> 
> Missing from the ingress info is the 'prefix' information for the path. We should display this, because it's quite important... since it can change the meaning of it quite a lot.

Made a field for it!

**Comment by vyncent-t** [comment]
> @vyncent-t as you suggested... How does wrapping work for the link? A good test for the storybook might be a very long host, and a very long path.

File: frontend/src/components/ingress/Details.tsx (hunk 5 above)
Comment (vyncent-t) [reply frontend/src/components/ingress/Details.tsx]: There is a link for the host field and for the path field!

File: frontend/src/components/ingress/Details.tsx [hunk 9]
```diff
@@ -36,6 +37,24 @@ export default function IngressDetails() {
     );
   }
 
+  /** this function is used to format the url to be used in the Link component */
+  function LinkStringFormat(url: any) {
+    try {
+      /** adding the protocol to the url string is required for the URL constructor to work because data.host is only the host name */
+      const formatURL = new URL(`http://${url}`);
```
[the remark this answers is not in the document]
Comment (vyncent-t) [reply frontend/src/components/ingress/Details.tsx]: Done!

**Comment by vyncent-t** [comment]
May need to see if multiple paths break any links... will look into this

**Comment by vyncent-t** [comment]
Can anyone provide context for this build error?
image

**Comment by vyncent-t** [comment]
Is there a way to fix this if I do not use mac?

**Comment by vyncent-t** [comment]
> May need to see if multiple paths break any links... will look into this

Fixed this, links will now print in a list.
image

**Comment by illume** [comment]
It would be nice to have a story for the states as well.

**Comment by vyncent-t** [comment]
WIP: this commit creates a working story using json dummydata

**Comment by vyncent-t** [comment]
Deleted old story file that was left in

**Comment by vyncent-t** [comment]
WIP
- Changed the backend field to list view
- Removed host from path link, links in path only render path strings but direct to full link
- Added context to path links

Need to remove comments and console logs before review

image

**Comment by vyncent-t** [comment]
Commit fixes issue with rendering links that are either "*" or ""

**Comment by vyncent-t** [comment]
Commit fixes issue with rendering links that are either "*" or ""

File: frontend/src/components/ingress/Details.tsx [hunk 10]
```diff
@@ -42,6 +43,110 @@ export default function IngressDetails() {
...
+  /** this function is used to determine the protocol used in the formatURL for LinkedStringFormat */
+  function getProtocol(item: any, url: any) {
+    const hostList: string[] = item.jsonData.spec?.tls?.map(({ ...hosts }) => `${hosts.hosts}`);
+    const isHttps = hostList.includes(`${url}`);
+
+    return isHttps;
+  }
+
+  /** this function is used to format the url to be used in the Link component */
+  function LinkStringFormat(url: any, item?: any, urlPath?: any) {
```
Comment (joaquimrocha) [inline frontend/src/components/ingress/Details.tsx:55]: The item is always used, as far as I can see, so it shouldn't be optional.
And the types should be precise.

File: frontend/src/components/ingress/Details.tsx [hunk 11]
```diff
@@ -42,6 +43,110 @@ export default function IngressDetails() {
     );
   }
 
+  /** this function is used to determine the protocol used in the formatURL for LinkedStringFormat */
+  function getProtocol(item: any, url: any) {
```
Comment (joaquimrocha) [inline frontend/src/components/ingress/Details.tsx:47]: This function is called `getProtocol` which suggests it returns the protocol, but it returns whether https is used.

File: frontend/src/components/ingress/Details.tsx [hunk 12]
```diff
@@ -42,6 +43,110 @@ export default function IngressDetails() {
...
+    if ((url === '*' || url === '' || url === undefined) && urlPath) {
+      return <Typography>{urlPath}</Typography>;
+    } else {
+      /* if the ingress does not use tls, then the url should be http */
+      if (!item.jsonData.spec?.tls) {
+        urlProtocol = 'http://';
+      } else {
+        /* if the ingress does use tls, then we compare the url to the hosts in the tls array to determine if the url should be http or https */
+        getProtocol(item, url) ? (urlProtocol = 'https://') : (urlProtocol = 'http://');
+      }
```
Comment (joaquimrocha) [inline frontend/src/components/ingress/Details.tsx:71]: This logic can be re-written much more clearly if the getProtocol function returns the actual protocol.

File: frontend/src/components/ingress/Details.tsx [hunk 13]
```diff
@@ -42,6 +43,110 @@ export default function IngressDetails() {
...
+  /** this function is used to format the url to be used in the Link component */
+  function LinkStringFormat(url: any, item?: any, urlPath?: any) {
+    let urlProtocol;
+
+    if ((url === '*' || url === '' || url === undefined) && !urlPath) {
+      return <Typography>{url}</Typography>;
+    }
+
+    if ((url === '*' || url === '' || url === undefined) && urlPath) {
+      return <Typography>{urlPath}</Typography>;
```
Comment (joaquimrocha) [inline frontend/src/components/ingress/Details.tsx:63]: Can't this be rewritten as:
```typescript
if (!url || url === '*') {
  return <Typography>{urlPath || url}</Typography>;
}
```
If url is undefined, we probably want to use it as `''` instead, so:
```typescript
if (!url || url === '*') {
  return <Typography>{urlPath || url || ''}</Typography>;
}
```

File: frontend/src/components/ingress/Details.tsx [hunk 14]
```diff
@@ -42,6 +43,110 @@ export default function IngressDetails() {
...
+      /* if the ingress does not use tls, then the url should be http */
+      if (!item.jsonData.spec?.tls) {
+        urlProtocol = 'http://';
+      } else {
+        /* if the ingress does use tls, then we compare the url to the hosts in the tls array to determine if the url should be http or https */
+        getProtocol(item, url) ? (urlProtocol = 'https://') : (urlProtocol = 'http://');
+      }
+
+      /* since we cannot access the prefix from the ingress object, we have to access it from the itemData array */
+      const itemData: any[] = item.jsonData.spec?.rules;
```
Comment (joaquimrocha) [inline frontend/src/components/ingress/Details.tsx:74]: You can actually use `item.spec.rules`. Because there's a spec getter in the class.
And the itemDatas should be called rules IMO.

File: frontend/src/components/ingress/Details.tsx [hunk 15]
```diff
@@ -42,6 +43,110 @@ export default function IngressDetails() {
...
+      let currentPathType;
+      if (itemData) {
+        itemData.forEach((rule: any) => {
+          rule.http.paths.forEach((path: any) => {
+            if (path.path === urlPath) {
+              currentPathType = path.pathType;
+            }
+          });
+        });
```
Comment (joaquimrocha) [inline frontend/src/components/ingress/Details.tsx:83]: I think using `for` loops here would be more efficient in not worse case scenarios, as you can break the loop once the path is found.

**Review by joaquimrocha — DISMISSED** [review]
Left a few comments in the code. Overall it's functionally fine but we can simplify the code a good deal.

File: frontend/src/components/ingress/Details.tsx (hunk 12 above)
Comment (vyncent-t) [reply frontend/src/components/ingress/Details.tsx:71]: I will change the name to reflect it being a Boolean return since it does not return a protocol object or string

File: frontend/src/components/ingress/Details.tsx (hunk 13 above)
Comment (vyncent-t) [reply frontend/src/components/ingress/Details.tsx:63]: I added notes to the relevant pieces for this to explain why one renders the host field and the other renders the path

File: frontend/src/components/ingress/Details.tsx [hunk 16]
```diff
@@ -42,6 +43,110 @@ export default function IngressDetails() {
...
+        }
+      } catch (e) {
+        console.debug("Ingress' host is not a valid URL; displaying it without a link.");
+        return url;
+      }
+    }
+  }
+
+  function backendFormat(backend: any) {
+    const backendList: any[] = [];
```
[the remark this answers is not in the document]
Comment (vyncent-t) [reply frontend/src/components/ingress/Details.tsx:117]: Ah right! I had reformatted the previous backend print logic to work without the label hover

File: frontend/src/components/ingress/Details.tsx [hunk 17]
```diff
@@ -42,6 +43,110 @@ export default function IngressDetails() {
...
+        itemData.forEach((rule: any) => {
+          rule.http.paths.forEach((path: any) => {
+            if (path.path === urlPath) {
+              currentPathType = path.pathType;
+            }
+          });
+        });
+      }
+
+      try {
```
[the remark this answers is not in the document]
Comment (vyncent-t) [reply frontend/src/components/ingress/Details.tsx:86]: Changed to be based on if else on the new URL creating a valid url or undefined

File: frontend/src/components/ingress/Details.tsx [hunk 18]
```diff
@@ -42,6 +43,113 @@ export default function IngressDetails() {
...
+  /** this function is used to determine the protocol used in the formatURL for LinkedStringFormat */
+  function isHttpsUsed(item: Ingress, url: String) {
+    const hostList: string[] = item.jsonData.spec?.tls?.map(({ ...hosts }) => `${hosts.hosts}`);
+    const isHttps = hostList.includes(`${url}`);
+
+    return isHttps;
+  }
+
+  /** this function is used to format the url to be used in the Link component */
+  function LinkStringFormat(url: String, item: Ingress, urlPath?: String) {
```
Comment (illume) [inline frontend/src/components/ingress/Details.tsx:55]: Components should take a `props`.

File: frontend/src/components/ingress/Details.tsx [hunk 19]
```diff
@@ -42,6 +43,113 @@ export default function IngressDetails() {
     );
   }
 
+  /** this function is used to determine the protocol used in the formatURL for LinkedStringFormat */
```
Comment (illume) [inline frontend/src/components/ingress/Details.tsx:46]: Other comments use a different format and style.

File: frontend/src/components/ingress/Details.tsx [hunk 20]
```diff
@@ -42,6 +43,113 @@ export default function IngressDetails() {
...
+  function isHttpsUsed(item: Ingress, url: String) {
+    const hostList: string[] = item.jsonData.spec?.tls?.map(({ ...hosts }) => `${hosts.hosts}`);
+    const isHttps = hostList.includes(`${url}`);
+
+    return isHttps;
+  }
+
+  /** this function is used to format the url to be used in the Link component */
```
Comment (illume) [inline frontend/src/components/ingress/Details.tsx:54]: Other functions use a different style of commenting.

File: frontend/src/components/ingress/Details.tsx [hunk 21]
```diff
@@ -42,6 +43,113 @@ export default function IngressDetails() {
...
+            {`(${currentPathType})`}
+          </Box>
+        );
+      } else if (formatURL) {
+        return <Link href={formatURL.toString()}>{`${formatURL.toString()}`}</Link>;
+      }
+    }
+  }
+
+  function backendFormat(backend: any) {
```
Comment (illume) [inline frontend/src/components/ingress/Details.tsx:119]: This one needs a comment.

**Review by illume — CHANGES_REQUESTED** [review]
Left a few notes.

File: frontend/src/components/ingress/Details.tsx (hunk 18 above)
Comment (illume) [inline frontend/src/components/ingress/Details.tsx:55]: How about adding a story for LinkStringFormat? You could add the different states, which would be a nice test.

File: frontend/src/components/ingress/Details.tsx (hunk 21 above)
Comment (illume) [inline frontend/src/components/ingress/Details.tsx:119]: This should be a React component (with a props argument).

**Comment by vyncent-t** [comment]
added stories for refactor

File: frontend/src/components/ingress/Details.tsx (hunk 18 above)
Comment (vyncent-t) [reply frontend/src/components/ingress/Details.tsx]: added!

File: frontend/src/components/ingress/Details.tsx (hunk 21 above)
Comment (vyncent-t) [reply frontend/src/components/ingress/Details.tsx]: done!

**Comment by vyncent-t** [comment]
Fixed multiple path story

File: frontend/src/components/ingress/Details.tsx [hunk 22]
```diff
@@ -1,12 +1,157 @@
...
 import Ingress, { IngressRule } from '../../lib/k8s/ingress';
+import { IngressBackend } from '../../lib/k8s/ingress';
 import { useSettings } from '../App/Settings/hook';
 import LabelListItem from '../common/LabelListItem';
 import { DetailsGrid } from '../common/Resource';
 import { SectionBox } from '../common/SectionBox';
 import SimpleTable from '../common/SimpleTable';
 
+/**
+ * This function is used to determine the protocol used in the formatURL for LinkedStringFormat
```
Comment (illume) [inline frontend/src/components/ingress/Details.tsx:13]: How about this?

```suggestion
Is https used in the ingress item?
```

"This function is used to" is a bit redundant.

File: frontend/src/components/ingress/Details.tsx [hunk 23]
```diff
@@ -1,12 +1,157 @@
...
+}
+
+export interface LinkStringFormatProps {
+  url: string;
+  item: Ingress;
+  urlPath?: string;
+}
+
+/**
+ * This function is used to format the url to be used in the Link component
```
Comment (illume) [inline frontend/src/components/ingress/Details.tsx:29]: Maybe get rid of this part: "This function is used to" which is a bit redundant for this component.

**Review by illume — APPROVED** [review]
Looks good to me. Thanks. 🎉

Nice addition of those tests.

I left a couple of minor nit picks about two comments for your consideration.

Feel free to merge it in when you're ready.
